### ISSUE TITLE
wrong positions for A1 mini 

### ISSUE THREAD

**Issue by Hillbilly-Phil** [eukatree/Bambu_CustomGCode#10]
Found some issues on the A1 mini code and patched them as good as I could. Runs smooth for me now:

-flushing takes place on wrong side of the bed --> added the right x coordinate
-extruder doesn't move far enough to the right to succesfully cut the filament --> added a slow move to cut the filament

Instructions for changing the filament with this modified gcode:

;===== Instructions: =====
; when print has paused, go into control
; set the nozzel temp to the temp you were printing with
; unload the filament by pushing on the upper extruder-button 
; load the new filament by pushing on the lower extruder-button 
; resume the print (flushing will happen next, flushing volumes can be set in Bambu Studio as if using an AMS)

[;===Bambu Lab A1 mini 0.4 GITHUB NO.txt](https://github.com/user-attachments/files/16149302/Bambu.Lab.A1.mini.0.4.GITHUB.NO.txt)

**Comment by pakonambawan** [comment]
i polished it right here. 
now after it cuts and goes back to it's place to poop, now it retracts the filament so just pull it out then just place another filament back in. 

[A1 mini MMS auto pause.txt](https://github.com/user-attachments/files/16457326/A1.mini.MMS.auto.pause.txt)

**Comment by khristangrey** [comment]
Will this work on the A1?

**Comment by pakonambawan** [comment]
> Will this work on the A1?

I haven't tried yet but I think no.  A1 mini have a smaller build plate so the gcode positions to where the cutter and poop ejector are different.

**Comment by khristangrey** [comment]
Got it. I want it to work on the A1. Can you point on what line I should change for the cutter positions and cutter? I have no experience on working on gcodes.

**Comment by Hillbilly-Phil** [comment]
Take a look at these lines:

[…]
G1 X0 Y90 F18000
G1 X-13.5 F9000
G1 E-13.5 F900

; pause for user to load and press resume
M400 U1
[…]

Meaning:
G1 = linear move to Position X Y Z with speed F
G1 E-… = Retract Filament by … mm
M400 U1 = Pauses the print till user resumes

So copy paste the whole code open in an text editor and play Around with the Coordinates X Y Z till you are satisfied with the positions. 
XYZ corespond to the xyz in the shown in the slicer. 
cheers

**Comment by Hillbilly-Phil** [comment]
Have a updated version with the toolbars cutting and retracting the filament wich I can upload as soon as I understand how to correctly commit on GitHub 😅 … and not just in the comment section.

**Comment by khristangrey** [comment]
I got it to work! I got the coordinates from the original gcode of Bambu A1. Thank you so much! 
[A1 No AMS.txt](https://github.com/user-attachments/files/16541185/A1.No.AMS.txt)

**Comment by Dennis-Q** [comment]
@Hillbilly-Phil  @pakonambawan and @eukatree  thank you for sharing all of this.

The instructions in the gcode was really useful (to not use the 'unload' / 'load' option). Also the finetuning done look great. I took the time to use your modifications and combine them with the latest version that is supplied with the latest Bambu Studio.

I have put your names in the gcode file and README.md since you are the contributors to the file as well.
The code on the github has been tested many times and works great for me with the latest Bambu Studio and A1 Mini firmware. Still, I am open to any suggestions for improvements if you have any.

My approach with the gcode file is to stay with the original code as much as possible and only change what is really mandatory to change (e.g. the gcode that causes the printer to crash). Also I tried to document what has been removed, added or moved, within the file. That could help later to understand the changes and why they were done. :)

You can find the code (for the A1 mini) here:
https://github.com/Dennis-Q/bambu

**Comment by wundergeu** [comment]
Thanks for your input concerning the necessary changes for the A1 (mini).

Out of curiosity:
Could I also just edit the 'Change Filament G-Code' section to a single line of 'M400 U1' and then load/unload the filament manually with the menu buttons on the display?

**Comment by Dennis-Q** [comment]
> Thanks for your input concerning the necessary changes for the A1 (mini).
> 
> Out of curiosity: Could I also just edit the 'Change Filament G-Code' section to a single line of 'M400 U1' and then load/unload the filament manually with the menu buttons on the display?

Nope, that's what I thought but things will not work as expected. One of the reasons is that an important part of the code is only executed when the AMS is connected and responding, which it is not.

Unload en load should not be used. Use Control and Extruder buttons as in the gcode instructions...

**Comment by wundergeu** [comment]
> Nope, that's what I thought but things will not work as expected. One of the reasons is that an important part of the code is only executed when the AMS is connected and responding, which it is not.
> 
> Unload en load should not be used. Use Control, Raise Temp and Extruder buttons as in the gcode instructions...

Alright, thanks for the quick response.
Gonna adapt the code for the A1. Your comments in the gcode make it way more understandable, so thanks for that.

**Comment by Joseluisuuj** [comment]
> ¡Lo hice funcionar! Obtuve las coordenadas del código G original de Bambu A1. ¡Muchas gracias! [A1 No AMS.txt](https://github.com/user-attachments/files/16541185/A1.No.AMS.txt)
hola, me he dado cuenta que la pausa la hace bien pero cuando se desplaza a la esquina a cortar el filamento no llega a cortarlo ya que no la presiona lo suficiende.
seria cosa de modificarlo para que avance mas y por ello la presione un poco mas y corte el filamento.
por lo demas el gcode para la bambulab A1 va bien

**Comment by Joseluisuuj** [comment]
> > ¡Lo hice funcionar! Obtuve las coordenadas del código G original de Bambu A1. ¡Muchas gracias! [A1 No AMS.txt](https://github.com/user-attachments/files/16541185/A1.No.AMS.txt)

hola, me he dado cuenta que la pausa la hace bien pero cuando se desplaza a la esquina a cortar el filamento no llega a cortarlo ya que no la presiona lo suficiende.
seria cosa de modificarlo para que avance mas y por ello la presione un poco mas y corte el filamento.
por lo demas el gcode para la bambulab A1 va bien

**Comment by Joseluisuuj** [comment]
#

> ¡Lo hice funcionar! Obtuve las coordenadas del código G original de Bambu A1. ¡Muchas gracias! [A1 No AMS.txt](https://github.com/user-attachments/files/16541185/A1.No.AMS.txt)

hola, me he dado cuenta que la pausa la hace bien pero cuando se desplaza a la esquina a cortar el filamento no llega a cortarlo ya que no la presiona lo suficiende.
seria cosa de modificarlo para que avance mas y por ello la presione un poco mas y corte el filamento.
por lo demas el gcode para la bambulab A1 va bien

**Comment by wundergeu** [comment]
> english: Hello, I have realized that the pause works well but when you move to the corner to cut the filament, it does not cut it because you do not press it hard enough.

Same for me. It doesn't seem to be "strong" enough to push the cutting lever. I don't know why it seems to work for others, but not for some.
I deleted the gcode for the cutting motion and just use the menu buttons to unload/load when it is paused.

**Comment by Joseluisuuj** [comment]
@khristangrey  Could you modify the gcode so that it would push the lever well and cut the filament?  thanks in advance

**Comment by khristangrey** [comment]
@Joseluisuuj Yes. I have not tried it again. But you are correct, it does not cut the filament. Will adjust it and will repost it here.

**Comment by Joseluisuuj** [comment]
Thank you very much, since that is the only thing I can print multicolor with

**Comment by Joseluisuuj** [comment]
@khristangrey you ajust the gcode??

**Comment by khristangrey** [comment]
So I tried to adjust the cutting motion but it won't work. The maximum movement I can do is 283 in this line:

[…]
G1 X256 F18000
G1 X283 F500 //here
G1 X256 F500
[…]

What happens if I add more, it grinds. Don't know why I can't make it to cut. For now, I would be manually pushing the cutting lever to cut the filament.

**Comment by mattmichaels** [comment]
perhaps these should be posted to the main repo?

i tried to print the repo last night and noticed the coordinates were not matching due to a different printer. would imagine others are having the same - didn't know the solution was hidden in open issue comments, haha.

**Comment by 1100101** [comment]
> So I tried to adjust the cutting motion but it won't work. The maximum movement I can do is 283 in this line:
> 
> […] G1 X256 F18000 G1 X283 F500 //here G1 X256 F500 […]
> 
> What happens if I add more, it grinds. Don't know why I can't make it to cut. For now, I would be manually pushing the cutting lever to cut the filament.

1. A1 mini users, be careful: the X movement numbers are for a A1, not the mini! This could damage the printer
2. Instead of increasing the X movement, I lowered the speed from 500 to 300. Doesn't work with all filaments, but better than before.

**Comment by mattmichaels** [comment]
i printed this on my A1 yesterday and had no cutting issues. 

once the pause started, i entered the control section and raised the temp of the extruder and extruded (up) to unload the current filament and then extruded (button down) to load the new filament. 

worked like a charm

**Comment by JimmyJo3D** [comment]
does anyone know if this gcode is the same one used here for a1?
https://makerworld.com/it/models/631637#profileId-556604
with this https://drive.google.com/file/d/172pAxlTttBCD0A5XI5EulBp2bkz6Kx5P/view

**Comment by 80prozent** [comment]
> I got it to work! I got the coordinates from the original gcode of Bambu A1. Thank you so much! [A1 No AMS.txt](https://github.com/user-attachments/files/16541185/A1.No.AMS.txt)

thanks for this code!

one thing i noticed in my first test-print is that it did not stop at the first filament change.

i think there is missing a `{endif} `

instead of:

```;===== machine: A1 mini edited in Juli 2024 =========================
;===== date: 20240518 =======================
G392 S0
M204 S9000
{if toolchange_count > 1}
G17
G2 Z{max_layer_z + 0.4} I0.86 J0.86 P1 F10000 ; spiral lift a little from second lift
G1 Z{max_layer_z + 3.0} F1200
```

it should be:

```;===== machine: A1 mini edited in Juli 2024 =========================
;===== date: 20240518 =======================
G392 S0
M204 S9000
{if toolchange_count > 1}
G17
G2 Z{max_layer_z + 0.4} I0.86 J0.86 P1 F10000 ; spiral lift a little from second lift
{endif}
G1 Z{max_layer_z + 3.0} F1200
```


i compared to https://drive.google.com/file/d/172pAxlTttBCD0A5XI5EulBp2bkz6Kx5P/view and this code indeed has the `{endif}` present.

**Comment by BencsikDani** [comment]
Hey Guys,

As I haven't really found a complete working "NoAMS + AutoUnload" code for the **A1 mini** that works, I took the original code from the Bambu Studio System presets, and I created my own from scratch. Of course I had lots of sources to get proper knowledge for it, like this original repo, other repos, and I also sprinkled with some AI magic.

I formatted & commented the original code, removed the AMS-related commands, and added Cut & Unload things.

With this implementation, the filament is really automatically unloaded, so when print is Paused, you just have to pull the filament out, feed the next one and hit Resume.

I hope it helps!

[a1_mini_change_filament_noams_autounload.txt](https://github.com/user-attachments/files/23597004/a1_mini_change_filament_noams_autounload.txt)

**UPDATE -> IT IS BAD, SEE V2 BELOW!**

**Comment by robinAZERTY** [comment]
> Hey Guys,
> 
> As I haven't really found a complete working "NoAMS + AutoUnload" code for the **A1 mini** that works, I took the original code from the Bambu Studio System presets, and I created my own from scratch. Of course I had lots of sources to get proper knowledge for it, like this original repo, other repos, and I also sprinkled with some AI magic.
> 
> I formatted & commented the original code, removed the AMS-related commands, and added Cut & Unload things.
> 
> With this implementation, the filament is really automatically unloaded, so when print is Paused, you just have to pull the filament out, feed the next one and hit Resume.
> 
> I hope it helps!
> 
> [a1_mini_change_filament_noams_autounload.txt](https://github.com/user-attachments/files/23597004/a1_mini_change_filament_noams_autounload.txt)


I had actually noticed that the code provided here:  
<https://github.com/eukatree/Bambu_CustomGCode/issues/10#issuecomment-2263087821>  
no longer works properly, likely due to recent changes in how certain G-code commands are interpreted. Bambu has updated both Bambu Studio and/or the printer firmware in recent versions, which can break scripts that previously worked.

On my side, I also had to update my own startup and filament change scripts:  
<https://github.com/robinAZERTY/BambuLab-A1-Mini-unofficial-profile>  
to make the behavior independent of any communication with the AMS.

However, as long as Bambu continues to adjust G-code properties or parsing, it's important to keep in mind that scripts like this may still change behavior — and could potentially become invalid or even unsafe depending on future updates.

**Comment by BencsikDani** [comment]
> Hey Guys,
> 
> As I haven't really found a complete working "NoAMS + AutoUnload" code for the **A1 mini** that works, I took the original code from the Bambu Studio System presets, and I created my own from scratch. Of course I had lots of sources to get proper knowledge for it, like this original repo, other repos, and I also sprinkled with some AI magic.
> 
> I formatted & commented the original code, removed the AMS-related commands, and added Cut & Unload things.
> 
> With this implementation, the filament is really automatically unloaded, so when print is Paused, you just have to pull the filament out, feed the next one and hit Resume.
> 
> I hope it helps!
> 
> [a1_mini_change_filament_noams_autounload.txt](https://github.com/user-attachments/files/23597004/a1_mini_change_filament_noams_autounload.txt)

Hey again,

After some weeks I finally realized how terrible my code was.
I am still figuring out some things, but until next time here is v2, that actually works.

[a1_mini_change_filament_noams_autounload_v2.txt](https://github.com/user-attachments/files/24043972/a1_mini_change_filament_noams_autounload_v2.txt)

Currently the flush amount during filament change is hardcoded. I found ~70-80 mm to be good enough. But feel free to change it in line 77.

I hope it helps!

**Comment by lucarosa** [comment]
> Hey again,
> 
> After some weeks I finally realized how terrible my code was. I am still figuring out some things, but until next time here is v2, that actually works.
> 
> [a1_mini_change_filament_noams_autounload_v2.txt](https://github.com/user-attachments/files/24043972/a1_mini_change_filament_noams_autounload_v2.txt)
> 
> Currently the flush amount during filament change is hardcoded. I found ~70-80 mm to be good enough. But feel free to change it in line 77.
> 
> I hope it helps!

This works perfectly! Thank you
